code/src/components/AddTodo.js

```diff
@@ -0,0 +1,81 @@
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import { useDispatch, useSelector } from 'react-redux';
+import { todos } from 'reducers/todos';
+import { visibilityFilter } from 'reducers/visibilityFilter';
+
+const StyledAddTodo = styled.div``;
+
+const Form = styled.form`
+  height: 55px;
+  display: grid;
+  grid-template-columns: 57px 12fr;
+  grid-gap: 0.5rem;
+  align-items: center;
+`;
+
+const Input = styled.input`
+  border: none;
+  font-size: 1rem;
+  margin-right: 0.5rem;
+
+  &::placeholder {
+    color: rgba(0, 0, 0, 0.3);
+  }
+
+  &:focus::placeholder {
+    color: #000;
+    transition: color 150ms ease-in-out;
+  }
+`;
+
+const Button = styled.button`
+  color: rgba(0, 0, 0, 0.3);
+  padding: 0 1rem 0 1.2rem;
+  background-color: transparent;
+  border: none;
+  font-weight: bold;
+  font-size: 1.5rem;
+
+  &:hover {
+    color: rgba(230, 0, 60, 1);
+    cursor: pointer;
+    transition: all 0.3s ease-in;
+  }
+`;
+
+export const AddTodo = () => {
+  const allTodos = useSelector(state => state.todos);
```

Comment:
Since you're only using the length of the array, I'd recommend returning that directly from the selector:

```
const numberOfTodos = useSelector(store => store.todos.length)
```

When you do it like you have, this `AddTodo ` component will re-render whenever the array of tasks changes (for example toggling a task's completion), but if you return the length from the selector, it'll only re-render when the length of the array changes. Small optimisation, but good to keep in the back of your head if you're working on a bigger project where you want to avoid unnecessary re-renders.

Reply:
I agree that the component will be re-rendered every time I add/delete a task since the length of the array will change. But I don't see how the length of the array will change by just toggling a task's completion. The number of tasks remain the same the way I see it.

When I look at the code again I'm not even sure why I'm using the length value. It must have been for something, but I can't remember what. Looking inside the styled component the empty property is not used anywhere.

```
<StyledAddTodo empty={allTodos.length === 0}>
````